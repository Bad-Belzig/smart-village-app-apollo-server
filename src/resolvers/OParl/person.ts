import { Body, IPerson, Location, Membership, Person } from '../../models';
import { findByIds, getPaginatedEntriesByIds } from '../resolverHelpers';

export const personResolvers = {
  Query: {
    oParlPersons: (_, args: { externalIds?: string[] }) =>
      args.externalIds ? findByIds(args.externalIds, Person) : Person.find(),
  },
  OParlPerson: {
    body: (parent: IPerson) => Body.findOne({ externalId: parent.body }),
    location: (parent: IPerson) =>
      Location.findOne({ externalId: parent.location }),
    membership: async (
      parent: IPerson,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Membership,
        parent.membership,
        args.offset,
        args.pageSize,
      ),
  },
};
