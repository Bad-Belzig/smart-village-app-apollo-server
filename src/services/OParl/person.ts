import { Person } from '../../models';
import { parsePerson } from '../../parser';
import { getJson, updateOrCreateEntry } from './importHelpers';
import { ImportQueue, ImportQueueEntry, ImportType } from './ImportTypes';

export const importPerson = async (value: unknown, queue?: ImportQueue) => {
  const json = await getJson(value);

  if (!json) return;

  const addToQueue: [ImportQueueEntry | ImportQueueEntry[], ImportType][] = [];

  if (json.body) {
    addToQueue.push([json.body, ImportType.Body]);
  }

  if (json.location) {
    addToQueue.push([json.location, ImportType.Location]);
  }

  if (json.membership) {
    addToQueue.push([json.membership, ImportType.Membership]);
  }

  try {
    return await updateOrCreateEntry(
      json,
      parsePerson,
      Person,
      addToQueue,
      queue,
    );
  } catch (e) {
    console.log(`Error while importing person from ${value}:`, e);
  }
};
