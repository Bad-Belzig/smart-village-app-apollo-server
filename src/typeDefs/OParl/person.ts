import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const personGQL = gql`
  extend type Query {
    oParlPersons(
      externalIds: [String!]
      keyword: [String!]
    ): [OParlPerson!]
  }

  type OParlPerson {
    affix: String
    body: OParlBody
    email: [String!]
    familyName: String
    formOfAddress: String
    gender: String
    givenName: String
    life: String
    lifeSource: String
    location: OParlLocation
    membership(offset: Int, pageSize: Int): [OParlMembership!]
    name: String
    phone: [String!]
    status: [String!]
    title: [String!]
    ${oParlBase}
  }
`;
