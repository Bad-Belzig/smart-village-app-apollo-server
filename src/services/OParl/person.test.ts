import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importPerson } from './person';

const helperFuncs = setupTestDataBase('test_person_import');

// TODO: change test url to something within our control
const testUrl =
  'https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/persons.asp?typ=sb&id=442';

describe('importing a person', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importPerson, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importPerson));

  afterAll(helperFuncs.afterAll);
});
