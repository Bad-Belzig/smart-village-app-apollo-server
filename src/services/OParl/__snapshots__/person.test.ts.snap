// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a person imports properly 1`] = `
Object {
  "__v": 0,
  "created": 2011-09-04T10:00:00.000Z,
  "email": Array [],
  "externalId": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/persons.asp?typ=sb&id=442",
  "familyName": "Hertel",
  "givenName": "Marita",
  "keyword": Array [],
  "membership": Array [],
  "modified": 2019-05-22T11:38:05.000Z,
  "name": "Marita Hertel",
  "phone": Array [],
  "status": Array [],
  "title": Array [],
  "type": "https://schema.oparl.org/1.0/Person",
}
`;
